pkg/cac/cac.go

```diff
@@ -2,25 +2,59 @@ package cac
 
 import (
 	"encoding/binary"
+	"errors"
 
 	"github.com/ethersphere/bee/pkg/bmtpool"
 	"github.com/ethersphere/bee/pkg/swarm"
 )
 
+var (
+	errTooShortChunkData = errors.New("short chunk data")
+	errTooLargeChunkData = errors.New("data too large")
+)
+
+// New creates a new content address chunk by initializing a span and appending the data to it.
 func New(data []byte) (swarm.Chunk, error) {
-	hasher := bmtpool.Get()
-	defer bmtpool.Put(hasher)
+	span := make([]byte, swarm.SpanSize)
+	binary.LittleEndian.PutUint64(span, uint64(len(data)))
+	return NewWithSpan(data, span)
+}
 
-	_, err := hasher.Write(data)
-	if err != nil {
-		return nil, err
+// NewWithDataSpan creates a new chunk assuming that the span precedes the actual data.
+func NewWithDataSpan(data []byte) (swarm.Chunk, error) {
+	if len(data) > swarm.ChunkSize+swarm.SpanSize {
+		return nil, errTooLargeChunkData
 	}
-	span := make([]byte, 8)
-	binary.LittleEndian.PutUint64(span, uint64(len(data)))
-	err = hasher.SetSpanBytes(span)
+
+	if len(data) < swarm.SpanSize {
+		return nil, errTooShortChunkData
+	}
+
+	return NewWithSpan(data[swarm.SpanSize:], data[:swarm.SpanSize])
+}
+
+// NewWithSpan creates a new chunk prepending the given span to the data.
+func NewWithSpan(data []byte, span []byte) (swarm.Chunk, error) {
+	h := hasher(data)
+	hash, err := h(span)
 	if err != nil {
 		return nil, err
 	}
+	return swarm.NewChunk(swarm.NewAddress(hash), append(span, data...)), nil
```

Comment:
This is a bit dangerous. Since `span` can actually be a slice which references storage that is not guaranteed to be immutable, you can potentially end up with a data race here, this is why the race detector complains. The problem is that you're reslicing data when calling this function from line 33. The slice you're shuffling around can actually cause all sorts of quirks. I would copy at least the span in `NewWithDataSpan` to a new slice, so that when you append `data` you actually append it to a new slice, preventing all sorts of peculiar bugs in the future.